neps/nep-9999.md

```diff
@@ -0,0 +1,111 @@
+---
+NEP: 9999
+Title: Parameter Weights
+Author: Andrei Kashin <andrei.kashin@near.org>, Jakob Meier <jakob@near.org>
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/9999
+Status: Draft
+Type: Protocol Track
+Category: Runtime
+Created: 26-Jan-2023
+---
+
+## Summary
+
+Allow to configure static weights for the gas costs of individual operations with which they are counted towards the chunk gas limit.
+
+## Motivation
+
+For NEAR blockchain stability, we need to ensure that blocks are produced regularly and in a timely manner.
+
+Gas chunk limit is used to ensure that a time to validate a chunk is strictly bounded by limiting the total cost of operations included in the chunk. This process relies on accurate estimates of gas costs for individual operations.
+
+Underestimating these costs leads to *undercharging* which can increase the chunk validation time and slow down the chunk production.
+
+As a concrete example, during the incremental deployment of Flat Storage we will be going through intermediate stages where the storage operations will be undercharged. Parameter Weights could be used to make such incremental deployment safer.
+
+## Rationale
+
+This design is easy to implement and simple to reason about and provides a clear way to address existing undercharging issues.
+
+If we don't do this work, it would be harder to deploy Flat Storage safely without a mechanism to safely adjust gas costs.
+
+## Alternatives
+
+### Increase the gas cost for undercharged operations
+We could change the gas costs for the operations that are undercharged to match the time it takes to process them.
+
+However, this would break contracts that rely on current gas costs. It can also steeply increase operating costs for the most active users of the blockchain. I believe the protocol must serve its users and not the other way around. Therefore, increasing gas costs without prior consent by the affected parties can only be considered in absolute emergency scenarios.
+
```

Comment:
We should also describe that:
- security related changes cannot be discussed in public and therefore we cannot always ask the smart contract devs for their opinion before we increase costs
- in case of the flat storage example, incrementing the cost to a safe value in the intermediate state would significantly increase the cost for many users, even though only very specific use cases are actually undercharged

Reply:
Good point about need to disclose the security-sensitive information, added it.

For the flat storage example, I don't think this problem is specific for this alternative solution, it's just the fact that our gas parameters are not granular/nuanced enough to charge two workloads differently according to their real resource usage. Parameter Weights would suffer from similar problem of limiting chunk sizes with some workloads that are actually fine from the processing time perspective just for the sake of curbing some worst-case workloads that we have in mind.

Reply:
I've added a paragraph dedicated to Flat Storage deployment, please take a look whether this expresses what you had in mind.